### ISSUE TITLE
Add Augmentation on Test Set

### ISSUE THREAD

**Issue by DimaMirana** [ultralytics/yolov5#2625]
## I want to train my model using the 
`!python train.py  --weights yolov5x.pt --data '../data.yaml' --epochs 5 --cache --img 416 --evolve`  
to generate the optimal hyperparameter value for augmentation and learning rate etc. Later I want to use that hyperparameter on the same model for my custom dataset. How do I add the augmentation parameter on train.py. In your tutorial on TTA(test time augmentation) just adding --augmentation is enough. Will it be same for the training set also? Also can I see total number of data generated after augmentation?
`!python train.py --batch 16 --weights yolov5x.pt --data '../data.yaml' --epochs 5 --cache --img 416 --hyp '/runs/train/evolve/hyp_evolved.yaml'` will this be the command for adding augmentation on the training set?

**Comment by glenn-jocher** [comment]
Augmentation does not generate new data, it simply views the existing data in different, random ways every time an image is used. Therefore an image is never seen twice the same way during training.
![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/112732849-c4f2fa80-8f3c-11eb-8a57-df1cee5c83fc.jpg)

Training augmentation settings are defined in a hyperparameter file. You can use hyperparameter evolution to optimize these values to your custom training requirements.
https://github.com/ultralytics/yolov5/blob/9b92d3ee769e8f26f2d535879dc69708998c47a3/data/hyp.scratch.yaml#L22-L33

Test time augmentation is not related to training augmentation. TTA simply runs inference at multiple image sizes and left-right flips. See TTA and Evolution tutorials for additional details:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://docs.ultralytics.com/yolov5/tutorials/model_export)
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by DimaMirana** [comment]
Is there any way to optimize only these values during evolution? And also after evolution `!python train.py --batch 16 --weights yolov5x.pt --data '../data.yaml' --epochs 5 --cache --img 416 --hyp '/runs/train/evolve/hyp_evolved.yaml`
This is the command to run augmentation on train data right?

**Comment by glenn-jocher** [comment]
@DimaMirana the meta dictionary in train.py controls hyperparameter constraints which you can use to freeze parameters.
https://github.com/ultralytics/yolov5/blob/9b92d3ee769e8f26f2d535879dc69708998c47a3/train.py#L543-L572

The command to run augmentation during training is:
```python train.py```

The hyp file used is defined here:
https://github.com/ultralytics/yolov5/blob/9b92d3ee769e8f26f2d535879dc69708998c47a3/train.py#L458

**Comment by DimaMirana** [comment]
I'm really sorry to disturb you but can you give me an example where you run augmentation during training with specific augmentation? Also if I just run
`python train.py` 
how will I can give the data folder path batch size etc.?

**Comment by glenn-jocher** [comment]
@DimaMirana training applies all augmentation hyperparameters by default. The hyp file used is defined here (point to another or modify this file if you'd like):
https://github.com/ultralytics/yolov5/blob/9b92d3ee769e8f26f2d535879dc69708998c47a3/train.py#L458

For a full list of training arguments and their defaults see train.py argparser:
https://github.com/ultralytics/yolov5/blob/ee169834bd0edf4e03b555688053da7bdd05a71e/train.py#L453-L489

or simply run `python train.py --help`
```bash
$ python train.py --help

usage: train.py [-h] [--weights WEIGHTS] [--cfg CFG] [--data DATA] [--hyp HYP] [--epochs EPOCHS] [--batch-size BATCH_SIZE] [--img-size IMG_SIZE [IMG_SIZE ...]] [--rect]
                [--resume [RESUME]] [--nosave] [--notest] [--noautoanchor] [--evolve] [--bucket BUCKET] [--cache-images] [--image-weights] [--device DEVICE] [--multi-scale]
                [--single-cls] [--adam] [--sync-bn] [--local_rank LOCAL_RANK] [--workers WORKERS] [--project PROJECT] [--entity ENTITY] [--name NAME] [--exist-ok] [--quad]
                [--linear-lr] [--upload_dataset] [--bbox_interval BBOX_INTERVAL] [--save_period SAVE_PERIOD] [--artifact_alias ARTIFACT_ALIAS]

optional arguments:
  -h, --help            show this help message and exit
  --weights WEIGHTS     initial weights path
  --cfg CFG             model.yaml path
  --data DATA           data.yaml path
  --hyp HYP             hyperparameters path
  --epochs EPOCHS
  --batch-size BATCH_SIZE
                        total batch size for all GPUs
  --img-size IMG_SIZE [IMG_SIZE ...]
                        [train, test] image sizes
  --rect                rectangular training
  --resume [RESUME]     resume most recent training
  --nosave              only save final checkpoint
  --notest              only test final epoch
  --noautoanchor        disable autoanchor check
  --evolve              evolve hyperparameters
  --bucket BUCKET       gsutil bucket
  --cache-images        cache images for faster training
  --image-weights       use weighted image selection for training
  --device DEVICE       cuda device, i.e. 0 or 0,1,2,3 or cpu
  --multi-scale         vary img-size +/- 50%
  --single-cls          train multi-class data as single-class
  --adam                use torch.optim.Adam() optimizer
  --sync-bn             use SyncBatchNorm, only available in DDP mode
  --local_rank LOCAL_RANK
                        DDP parameter, do not modify
  --workers WORKERS     maximum number of dataloader workers
  --project PROJECT     save to project/name
  --entity ENTITY       W&B entity
  --name NAME           save to project/name
  --exist-ok            existing project/name ok, do not increment
  --quad                quad dataloader
  --linear-lr           linear LR
  --upload_dataset      Upload dataset as W&B artifact table
  --bbox_interval BBOX_INTERVAL
                        Set bounding-box image logging interval for W&B
  --save_period SAVE_PERIOD
                        Log model after every "save_period" epoch
  --artifact_alias ARTIFACT_ALIAS
                        version of dataset artifact to be used

```

**Comment by DimaMirana** [comment]
So if I want to run training on my custom data .yaml for 5 epoch with my finding hyperparameter after training then this will be the command?
`!python train.py --batch 16 --weights yolov5x.pt --data '../data.yaml' --epochs 5 --cache --img 416 --hyp '/runs/train/evolve/hyp_evolved.yaml'`
Also can you tell me --epoch 5 in --evolve mode mean? Does it mean that this model run for 5 epoch and on each epoch it evolves the best model 300 times to get the optimal hyperparams?
`!python train.py --weights yolov5x.pt --data '../data.yaml' --epochs 5 --cache --img 416 --evolve`

**Comment by glenn-jocher** [comment]
@DimaMirana see Hyperparameter Evolution Tutorial for complete details on evolution. Evolution and training are seperate. Evolution will return a set of hyps that you can then point to to train, i.e. `python train.py --hyp your_hyps.yaml`.

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [ONNX and TorchScript Export](https://docs.ultralytics.com/yolov5/tutorials/model_export)
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by DimaMirana** [comment]
got it thanks. Also how to cite your project in papers? Where is the papaer?

**Comment by DimaMirana** [comment]
Also where can I see the overall Accuracy of the model of all classes. not the precision and recall.

**Comment by DimaMirana** [comment]
also any image about the yolov5 network architecture?

**Comment by glenn-jocher** [comment]
@DimaMirana please see https://github.com/ultralytics/yolov5#citation for citations. You can use netron or tensorboard to visualize architecture, and also search issues for custom visualizations.

**Comment by DimaMirana** [comment]
how to visualize architecture using tensorboard?

**Comment by glenn-jocher** [comment]
@DimaMirana in theory you should be able to uncomment L335 in train.py to visualize model with tensorboard, though we've had various various levels of success in practice.
https://github.com/ultralytics/yolov5/blob/9ccfa85249a2409d311bdf2e817f99377e135091/train.py#L329-L335

**Comment by DimaMirana** [comment]
Thanks for the info. Also where can I found the f1 score of each class after finishing training? the current metrrics gives the following.
 Class      Images      Labels           P           R      mAP@.5  mAP@.5:.95:

**Comment by glenn-jocher** [comment]
@DimaMirana F1 curves are computed and displayed in F1.jpg in your results directory, though it is not printed to screen. VOC example with YOLOv5x6:
https://wandb.ai/glenn-jocher/VOC

![image](https://user-images.githubusercontent.com/26833433/113564351-8a0e5800-9609-11eb-8165-3e03e0ffa0e0.png)

**Comment by DimaMirana** [comment]
is there any way to get F1 score of each class?

**Comment by DimaMirana** [comment]
Also when using colab is there any way to save the weights and other necessary files so that i can resume it from previous training on the next day?

**Comment by glenn-jocher** [comment]
> is there any way to get F1 score of each class?

It seems like the best way to get F1 at each class would be to add it to the printed metrics in test.py here. You should note though that each class also has a highest F1 at a different confidence.
https://github.com/ultralytics/yolov5/blob/9ccfa85249a2409d311bdf2e817f99377e135091/test.py#L227-L230

**Comment by DimaMirana** [comment]
Thanks. Also when using colab is there any way to save the weights and other necessary files so that i can resume it from previous training on the next day?

**Comment by glenn-jocher** [comment]
@DimaMirana yes there's two tricks. You can set your --project directory to a mounted Google Drive, or you can also try the new W&B integration which should save your checkpoints to W&B now also. See https://wandb.ai/cayush/yolov5-dsviz-demo/reports/Object-Detection-with-YOLO-and-Weights-Biases--Vmlldzo0NTgzMjk

**Comment by DimaMirana** [comment]
Many thanks.
